include/zephyr/mgmt/mcumgr/serial.h

```diff
@@ -129,11 +130,12 @@ struct net_buf *mcumgr_serial_process_frag(
  * @param len                   The length of the unencoded mcumgr packet.
  * @param cb                    A callback used to transmit raw bytes.
  * @param arg                   An optional argument to pass to the callback.
+ * @param tx_data               Overall packet context
  *
  * @return                      0 on success; negative error code on failure.
  */
-int mcumgr_serial_tx_pkt(const uint8_t *data, int len, mcumgr_serial_tx_cb cb,
-			 void *arg);
+int mcumgr_serial_tx_pkt(const uint8_t *data, int len, mcumgr_serial_tx_cb cb, void *arg,
+			 struct zephyr_tx_information *tx_data);
```

Comment:
The length and crc are appended at the bottom level, at mcumgr_serial_tx_frame in serial_util.c, that means it has to transverse all of these layers:
```
#0  mcumgr_serial_tx_frame (data=data@entry=0x200064c0 <net_buf_data_pkt_pool+768> "\003", len=len@entry=256, 
    cb=cb@entry=0x1aca5 <smp_shell_tx_raw>, arg=arg@entry=0x0, tx_data=tx_data@entry=0x20005ad0 <sys_work_q_stack+1944>, 
    out_data_bytes_txed=out_data_bytes_txed@entry=0x20005a94 <sys_work_q_stack+1884>)
    at zephyr/subsys/mgmt/mcumgr/serial_util.c:198
#1  0x0001b1fc in mcumgr_serial_tx_pkt (data=0x200064c0 <net_buf_data_pkt_pool+768> "\003", len=256, 
    cb=cb@entry=0x1aca5 <smp_shell_tx_raw>, arg=arg@entry=0x0, tx_data=0x20005ad0 <sys_work_q_stack+1944>)
    at zephyr/subsys/mgmt/mcumgr/serial_util.c:323
#2  0x000126ea in smp_shell_tx_pkt (zst=<optimized out>, nb=0x20006190 <_net_buf_pkt_pool+48>, tx_data=<optimized out>)
    at zephyr/subsys/mgmt/mcumgr/smp_shell.c:183
#3  0x0001ac68 in zephyr_smp_tx_rsp (ns=<optimized out>, rsp=<optimized out>, arg=0x200015f0 <smp_shell_transport>)
    at zephyr/subsys/mgmt/mcumgr/smp.c:185
#4  0x000137f6 in smp_process_request_packet (streamer=streamer@entry=0x20005b34 <sys_work_q_stack+2044>, 
    vreq=0x20006160 <_net_buf_pkt_pool>) at zephyr/subsys/mgmt/mcumgr/lib/smp/src/smp.c:297
#5  0x00012694 in zephyr_smp_process_packet (nb=<optimized out>, zst=0x200015f0 <smp_shell_transport>)
    at zephyr/subsys/mgmt/mcumgr/smp.c:218
```
zephyr_smp_tx_rsp is the function where the packet starts getting chopped up, at the start of the function, all the data is available for knowing the full length and checksum of the packet, after 1 iteration the buffer is missing the complete set of data

Reply:
My bad. Anyway: dump the `zephyr_smp_tx_rsp` and move fragmentation directly to transports.
Why? Because all transports directly access `net_buf->data` so they can loop themselves over the data with the mtu they have.
This helps us with a few problems:
 1) exhausting mcumgr `net_buf` pool for fragmenting packets for transports with mtu smaller than packet, as currently buffers for "fragments" are taken from the mcumgr pool;
 2) having two levels of fragmentation, for example when uart mtu does not fit into uart encoding buffer you will have to fragment the output twice; this can be done once directly in uart where you select smaller of the values (mtu, buffer size) and fragment buffer once.
 3) we do not have to pass mtu to upper levels - one callback less.

Reply:
Will have to redo this review but will need the code in this version of the review to be updated on zephyr 2.7 since this bug is present there

Reply:
...Should probably have tested with 2.7 sooner, turns out 2.7 doesn't have this issue, will rework this review.

Reply:
Fixed